tests/Unit/Pypp/PyppFundamentals.hpp

```diff
@@ -432,50 +420,64 @@ T* get_ptr_to_elem(PyObj* npy_array, const std::array<size_t, Size>& idx) {
 }
 
 template <typename T>
-struct FromPyObject<T, Requires<tt::is_a_v<Tensor, T> and T::rank() != 0 and
-                                cpp17::is_same_v<typename T::type, double>>> {
-  static T convert(PyObject* p) {
-    if (p == nullptr) {
-      throw std::runtime_error{"Received null PyObject."};
-    }
-    // clang-tidy: c-style casts. (Expanded from macro)
-    if (not PyArray_CheckExact(p)) {  // NOLINT
-      throw std::runtime_error{"Cannot convert non-array type to Tensor."};
-    }
-    // clang-tidy: reinterpret_cast
-    const auto npy_array = reinterpret_cast<PyArrayObject*>(p);  // NOLINT
-    if (PyArray_TYPE(npy_array) != NPY_DOUBLE) {
-      throw std::runtime_error{
-          "Cannot convert array of non-double type to Tensor."};
-    } else if (PyArray_NDIM(npy_array) != static_cast<long>(T::rank())) {
+T tensor_conversion_impl(PyObject* p) {
+  if (p == nullptr) {
+    throw std::runtime_error{"Received null PyObject."};
+  }
+  // clang-tidy: c-style casts. (Expanded from macro)
+  if (not PyArray_CheckExact(p)) {  // NOLINT
+    throw std::runtime_error{"Cannot convert non-array type to Tensor."};
+  }
+  // clang-tidy: reinterpret_cast
+  const auto npy_array = reinterpret_cast<PyArrayObject*>(p);  // NOLINT
+  if (PyArray_TYPE(npy_array) != NPY_DOUBLE) {
+    throw std::runtime_error{
+        "Cannot convert array of non-double type to Tensor."};
+  } else if (PyArray_NDIM(npy_array) != static_cast<long>(T::rank())) {
+    throw std::runtime_error{
+        "Mismatch between ndim of numpy ndarray and rank of Tensor."};
+  }
+
+  const auto npy_array_dims = PyArray_DIMS(npy_array);
+  const auto& t_array_dims = T::index_dims();
```

Comment:
my guess is that this could be `constexpr auto`

Reply:
@osheamonn did you see this comment?